### ISSUE TITLE
ECS Agent unable to start

### ISSUE THREAD

**Issue by rahearn** [aws/amazon-ecs-agent#270]
Every instance we've tried to stand up today has failed with the following log messages repeated over and over.

```
2015-12-15T20:19:49Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.us-east-1.amazonaws.com/ module="tcs handler"
2015-12-15T20:19:49Z [INFO] Creating poll dialer module="ws client" host="ecs-t-2.us-east-1.amazonaws.com"
2015-12-15T20:19:52Z [ERROR] Error connecting to TCS: dial tcp: i/o timeout module="tcs handler"
2015-12-15T20:19:52Z [INFO] Error from tcs; backing off module="tcs handler" err="dial tcp: i/o timeout"
2015-12-15T20:20:11Z [DEBUG] Connecting to ACS endpoint https://ecs-a-2.us-east-1.amazonaws.com/ module="acs handler"
2015-12-15T20:20:11Z [DEBUG] Added update handlers module="updater"
2015-12-15T20:20:11Z [INFO] Creating poll dialer module="ws client" host="ecs-a-2.us-east-1.amazonaws.com"
2015-12-15T20:20:14Z [ERROR] Error connecting to ACS: dial tcp: i/o timeout module="acs handler"
2015-12-15T20:20:14Z [INFO] Error from acs; backing off module="acs handler" err="dial tcp: i/o timeout"
```

We're using the ECS optimized AMI 2015.09.b (ami-ddc7b6b7)

**Comment by djenriquez** [comment]
Which version of ecs-agent?

**Comment by rahearn** [comment]
1.7.0

**Comment by rahearn** [comment]
I seem to have found the trigger.

This will happen if the first `nameserver` in /etc/resolve.conf doesn't respond, even if a later `nameserver` is working fine. If the first server responds to dns queries, but can't resolve the `*.amazonaws.com` queries it will still work fine.

**Comment by euank** [comment]
I'm glad you figured out what was causing this.

What's interesting here is that the HTTP client used for the first several API calls worked and the websocket connection didn't.

It's possible that the difference in timeouts is what's causing this (30s vs 3s I believe). It's also possible there's some other difference between the two network dials for those two connections.
If you happen to have Agent logs that include the "Registered!" logline on an affected instance, it would be interesting to see the timestamp for that and previous lines.

What do you think the right behavior here should be? It seems to me like this might be working-as-expected other than perhaps making those timeouts the same.

Thanks,
Euan

**Comment by rahearn** [comment]
Here's the full logs for an affected instance from start up to the first failure.

```
2015-12-18T20:10:44Z [INFO] Starting Agent: Amazon ECS Agent - v1.7.0 (191dbd5)
2015-12-18T20:10:44Z [INFO] Loading configuration
2015-12-18T20:10:44Z [DEBUG] Loaded config: Cluster: , Region: us-east-1, DataDir: /data, Checkpoint: true, AuthType: dockercfg, UpdatesEnabled: true, DisableMetrics: false, ReservedMem: 0
2015-12-18T20:10:44Z [INFO] Checkpointing is enabled. Attempting to load state
2015-12-18T20:10:44Z [INFO] Loading state! module="statemanager"
2015-12-18T20:10:44Z [DEBUG] Loaded state! module="statemanager" state="&{Data:map[Cluster:0xc208099090 EC2InstanceID:0xc2080990b0 ACSSeqNum:0xc2080990d0 TaskEngine:0xc208099050 ContainerInstanceArn:0xc208099070] Version:4}"
2015-12-18T20:10:44Z [DEBUG] Setting cluster to default; none configured
2015-12-18T20:10:44Z [INFO] Restored cluster 'default'
2015-12-18T20:10:44Z [INFO] Detected Docker versions [1.17 1.18 1.19 1.20]
2015-12-18T20:10:44Z [INFO] Restored from checkpoint file. I am running as 'arn:aws:ecs:us-east-1:714955947340:container-instance/7f2d7c78-f720-4d40-8b4b-40d859c34030' in cluster 'default'
2015-12-18T20:10:54Z [INFO] Registered! module="api client"
2015-12-18T20:10:54Z [INFO] Saving state! module="statemanager"
2015-12-18T20:10:54Z [INFO] Beginning Polling for updates
2015-12-18T20:10:54Z [INFO] Initializing stats engine module="stats"
2015-12-18T20:10:54Z [DEBUG] Could not map container to task, ignoring module="stats" err="Could not map docker id to task" id="075f5ed5c5e0ca09cbc28834dd67e1311ebd067d1c20fc350f8be53192594874"
2015-12-18T20:10:54Z [DEBUG] Connecting to ACS endpoint https://ecs-a-2.us-east-1.amazonaws.com/ module="acs handler"
2015-12-18T20:10:54Z [DEBUG] Added update handlers module="updater"
2015-12-18T20:10:54Z [INFO] Creating poll dialer module="ws client" host="ecs-a-2.us-east-1.amazonaws.com"
2015-12-18T20:10:54Z [DEBUG] Connecting to TCS endpoint https://ecs-t-2.us-east-1.amazonaws.com/ module="tcs handler"
2015-12-18T20:10:54Z [INFO] Creating poll dialer module="ws client" host="ecs-t-2.us-east-1.amazonaws.com"
2015-12-18T20:10:57Z [ERROR] Error connecting to ACS: dial tcp: i/o timeout module="acs handler"
2015-12-18T20:10:57Z [INFO] Error from acs; backing off module="acs handler" err="dial tcp: i/o timeout"
2015-12-18T20:10:57Z [ERROR] Error connecting to TCS: dial tcp: i/o timeout module="tcs handler"
2015-12-18T20:10:57Z [INFO] Error from tcs; backing off module="tcs handler" err="dial tcp: i/o timeout"
```

I would expect that the ecs-agent would be able to operate properly as long as any `nameserver` is working. That's part of the point of configuring more than one `nameserver` in the first place. The non-response nameserver would have been much easier to diagnose if I had been able to deploy a task to this instance that was relying on the dead nameserver to resolve private dns entries (name not resolved errors instead of i/o timeout error).

**Comment by asans** [comment]
I am experiencing similar issue.  I'm running ecs-agent on CoreOS.  Here's the interesting tidbit:

* I have consul agent running on CoreOS that is registered as an additional nameserver in the resolv.conf file.  This consul agent's DNS is for internal service resolution.  Additionally, I've added a recursor option so that any non-consul related queries are forwarded to the normal DNS server.

* This is hosted on a AWS instance inside a VPC. 

* Upon running the agent, I get pretty much the same log as the one above.  API client registers properly but the ws clients do not register and errors out with i/o timeouts.  

```
2016-01-07T17:39:59Z [INFO] Starting Agent: Amazon ECS Agent - v1.7.0 (191dbd5)
326]: 2016-01-07T17:39:59Z [INFO] Loading configuration
326]: 2016-01-07T17:39:59Z [INFO] Checkpointing not enabled; a new container instance will be created each time the agent is run
326]: 2016-01-07T17:39:59Z [INFO] Detected Docker versions [1.17 1.18 1.19 1.20]
326]: 2016-01-07T17:39:59Z [INFO] Registering Instance with ECS
326]: 2016-01-07T17:40:09Z [INFO] Registered! module="api client"
326]: 2016-01-07T17:40:09Z [INFO] Registration completed successfully. I am running as 'arn:aws:ecs:us-west-2:952153266875:container-ins
326]: 2016-01-07T17:40:09Z [INFO] Beginning Polling for updates
326]: 2016-01-07T17:40:09Z [INFO] Initializing stats engine module="stats"
326]: 2016-01-07T17:40:09Z [INFO] Creating poll dialer module="ws client" host="ecs-a-1.us-west-2.amazonaws.com"
326]: 2016-01-07T17:40:09Z [INFO] Creating poll dialer module="ws client" host="ecs-t-1.us-west-2.amazonaws.com"
326]: 2016-01-07T17:40:12Z [ERROR] Error connecting to ACS: dial tcp: i/o timeout module="acs handler"
326]: 2016-01-07T17:40:12Z [INFO] Error from acs; backing off module="acs handler" err="dial tcp: i/o timeout"
326]: 2016-01-07T17:40:12Z [ERROR] Error connecting to TCS: dial tcp: i/o timeout module="tcs handler"
326]: 2016-01-07T17:40:12Z [INFO] Error from tcs; backing off module="tcs handler" err="dial tcp: i/o timeout"
326]: 2016-01-07T17:40:12Z [INFO] Creating poll dialer module="ws client" host="ecs-a-1.us-west-2.amazonaws.com"
326]: 2016-01-07T17:40:13Z [INFO] Creating poll dialer module="ws client" host="ecs-t-1.us-west-2.amazonaws.com"
326]: 2016-01-07T17:40:15Z [ERROR] Error connecting to ACS: dial tcp: i/o timeout module="acs handler"
326]: 2016-01-07T17:40:15Z [INFO] Error from acs; backing off module="acs handler" err="dial tcp: i/o timeout"
```

* Upon removing consul dns from the resolv.conf file, amazon ecs-agent is able to start up and properly connect with the websockets.  Everything looks good.

doing a dig on both the consul dns and the normal aws dns both return IPs properly.  

So what I don't get here is why amazon ecs-agent is having such difficulty trying to connect when api client obviously connects without problem?

**Comment by djenriquez** [comment]
These logs are specific to the issue but not detailed to point to a resolution. I've had these same error logs and the fix for me was enabling the correct API access.

AWS has a policy for instances running the ecs-agent. Assuming that your instance is running a role, if you attach the "AmazonEC2ContainerServiceEC2Role" policy to that role, it might just solve your issue. It did for me atleast. After applying the policy, restart the container and give it a few minutes for it to realize it now has access.

**Comment by euank** [comment]
Thanks for the additional information @asans 

The reason that the websocket vs api connection differs is that they have a different timeout (30 seconds vs 3 seconds). The logs show it takes 10 seconds from attempting to register to successfully doing so and, presumably, the websocket connection would also take 10 seconds and thus times out.

A quick workaround to at least make this consistent is to make them both have the same  30s timeout.

That doesn't address what I see as the bigger problem, however. DNS resolution shouldn't take 10s.

The next points of investigation I think are:
1. Is the Agent's `resolv.conf` (visible via `docker cp`) different than the host's?
2. Can this failure be reproduced with dig from within the Agent's namespace?
3. Does the difference between the `netgo` stack the Agent uses vs the `libc` dns functions cause this?

**Comment by asans** [comment]
It is working when consul dns is removed from resolv.conf, so it's not the role policy as that's already been set and is made to work.  

After further trial and testing, I was finally able to get this to work.  When it comes to using Consul DNS servers for resolution, one must expose both the TCP and UDP ports to connect to it.  So `-p 53:8600` and `-p 53:8600/udp` are required.  In most of the docker container launch code out there on the Internet, consul is only started up without opening the TCP port. 

Since I was able to get this to work, this means that the DNS query for host ecs-a-1.us-west-2.amazonaws.com probably exceeded the DNS UDP Packet limit, so TCP was used instead to connect.  Since TCP was not exposed, i/o timeout occurred.  

Hopefully this will help those who use Consul DNS servers running inside a docker container.

**Comment by samuelkarp** [comment]
We've bumped the timeout up as a mitigation and to make the different http clients behave in the same way.  This will be part of our next release.

**Comment by samuelkarp** [comment]
Released with [1.7.1](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.7.1).
